lib/ruby_smb/gss/provider/ntlm.rb

```diff
@@ -0,0 +1,337 @@
+          # which is then put back into a new GSS reply-blob
+          def process_gss_type1(gss_api)
+            unless Gss.asn1dig(gss_api, 1, 0, 0, 0, 0)&.value == Gss::OID_NTLMSSP.value
+              return
+            end
+
+            raw_type1_msg = Gss.asn1dig(gss_api, 1, 0, 1, 0)&.value
+            return unless raw_type1_msg
+
+            type1_msg = Net::NTLM::Message.parse(raw_type1_msg)
+            if type1_msg.flag & NEGOTIATE_FLAGS[:UNICODE] == NEGOTIATE_FLAGS[:UNICODE]
+              type1_msg.domain.force_encoding('UTF-16LE')
+              type1_msg.workstation.force_encoding('UTF-16LE')
+            end
+            type2_msg = process_ntlm_type1(type1_msg)
+
+            Result.new(Gss.gss_type2(type2_msg.serialize), WindowsError::NTStatus::STATUS_MORE_PROCESSING_REQUIRED)
+          end
+
+          # take the GSS blob, extract the NTLM type 3 message and pass it to the process method to build the response
+          # which is then put back into a new GSS reply-blob
+          def process_gss_type3(gss_api)
+            neg_token_init = Hash[RubySMB::Gss.asn1dig(gss_api, 0).value.map { |obj| [obj.tag, obj.value[0].value] }]
+            raw_type3_msg = neg_token_init[2]
+
+            type3_msg = Net::NTLM::Message.parse(raw_type3_msg)
+            if type3_msg.flag & NEGOTIATE_FLAGS[:UNICODE] == NEGOTIATE_FLAGS[:UNICODE]
+              type3_msg.domain.force_encoding('UTF-16LE')
+              type3_msg.user.force_encoding('UTF-16LE')
+              type3_msg.workstation.force_encoding('UTF-16LE')
+            end
+
+            nt_status = process_ntlm_type3(type3_msg)
+            buffer = identity = nil
+
+            case nt_status
+            when WindowsError::NTStatus::STATUS_SUCCESS
+              buffer = OpenSSL::ASN1::ASN1Data.new([
+                OpenSSL::ASN1::Sequence.new([
+                  OpenSSL::ASN1::ASN1Data.new([
+                    OpenSSL::ASN1::Enumerated.new(OpenSSL::BN.new(0)),
+                  ], 0, :CONTEXT_SPECIFIC)
+                ])
+              ], 1, :CONTEXT_SPECIFIC).to_der
+
+              account = @provider.get_account(
+                type3_msg.user,
+                domain: type3_msg.domain
+              )
+              if account.nil?
+                if @provider.allow_anonymous
+                  identity = IDENTITY_ANONYMOUS
+                end
+              else
+                identity = account.to_s
+              end
+            end
+
+            Result.new(buffer, nt_status, identity)
+          end
+        end
+
+        # @param [Boolean] allow_anonymous whether or not to allow anonymous authentication attempts
+        # @param [String] default_domain the default domain to use for authentication, unless specified 'WORKGROUP' will
+        #   be used
+        def initialize(allow_anonymous: false, default_domain: nil)
+          @allow_anonymous = allow_anonymous
+          @default_domain = default_domain || 'WORKGROUP'
```

Comment:
```suggestion
        def initialize(allow_anonymous: false, default_domain: 'WORKGROUP')
          @allow_anonymous = allow_anonymous
          @default_domain = default_domain
```

Would this do the same thing?

Reply:
Doing it like `def initialize(allow_anonymous: false, default_domain: 'WORKGROUP')` would allow the `default_domain` to be nil if someone explicitly set it to nil which would cause errors later on. I was trying to avoid that since a default domain is necessary. I guess I could raise an arg error if the domain is invalid. That might be more clear.

Reply:
Yeah an arg error would make it clearer, especially if someone is just jumping into the method definition to quickly check what is and isn't allowed